<a href="https://colab.research.google.com/github/JulieWang417/DS-Unit-1-Sprint-2-Statistics/blob/master/module3/LS_DS_123_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 123

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate):` 

You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).



### （1）Write a function：prob_drunk_given_positive

In [0]:
P_drunkdriver = 0.001
P_non_drunkdriver = 1-P_drunkdriver
P_pos_drunkdriver = 1
P_pos_non_drunkdriver = 0.08


In [0]:
def prob_drunk_given_positive(P_drunkdriver, P_pos_non_drunkdriver):
  P_non_drunkdriver = 1-P_drunkdriver
  P_pos_drunkdriver = 1
  posterior_prob = (P_pos_drunkdriver*P_drunkdriver)/(P_pos_drunkdriver*P_drunkdriver+P_pos_non_drunkdriver*P_non_drunkdriver)
  return posterior_prob

What is the probability that a person is drunk after one positive breathalyzer test?

In [11]:
posterior_prob_first = prob_drunk_given_positive(0.001,0.08)
posterior_prob_first

0.012357884330202669

What is the probability that a person is drunk after two positive breathalyzer tests?

In [12]:
posterior_prob_second = prob_drunk_given_positive(0.012358,0.08)
posterior_prob_second

0.135253218365544

How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

In [13]:
posterior_prob_third = prob_drunk_given_positive(0.13525,0.08)
posterior_prob_third

0.6615956562148413

In [14]:
posterior_prob_forth = prob_drunk_given_positive(0.661596,0.08)
posterior_prob_forth

0.9606888843093581

We have a probabiility 0.96 after the forth time we did the positive breathalyzer tests.

### (3) In your own words, summarize the difference between Bayesian and Frequentist statistics

Both of Bayesian and Frequentist approaches allow one to evaluate evidence about competing hypotheses.

**Bayesian inference:**

• uses probabilities for both hypotheses and data.

• depends on the prior and likelihood of observed data.

• requires one to know or construct a ‘subjective prior’.

• dominated statistical practice before the 20th century.

• may be computationally intensive due to integration over many parameters.

**Frequentist inference:**

• never uses or gives the probability of a hypothesis (no prior or posterior).

• depends on the likelihood P(D | H)) for both observed and unobserved data.

• does not require a prior.

• dominated statistical practice during the 20th century.

• tends to be less computationally intensive.


Frequentist measures like p-values and confidence intervals continue to dominate research,
especially in the life sciences. However, in the current era of powerful computers and
big data, Bayesian methods have undergone an enormous renaissance in fields like machine learning and genetics. There are now a number of large, ongoing clinical trials using
Bayesian protocols, something that would have been hard to imagine a generation ago.
While professional divisions remain, the consensus forming among top statisticians is that
the most effective approaches to complex problems often draw on the best insights from
both schools working in concert.

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!